icu4c/source/i18n/formatted_string_builder.cpp

```diff
@@ -333,16 +360,23 @@ int32_t FormattedStringBuilder::prepareForInsertHelper(int32_t index, int32_t co
                 sizeof(Field) * (fLength - index));
 
         fZero = newZero;
-        fLength += count;
+        fLength = newLength;
+    }
+    int32_t result;
+    if (uprv_add32_overflow(fZero, index, &result)) {
+        status = U_INPUT_TOO_LONG_ERROR;
+        return -1;
     }
-    U_ASSERT((fZero + index) >= 0);
-    return fZero + index;
+    return result;
 }
 
-int32_t FormattedStringBuilder::remove(int32_t index, int32_t count) {
+int32_t FormattedStringBuilder::remove(int32_t index, int32_t count, UErrorCode &status) {
     // TODO: Reset the heap here?  (If the string after removal can fit on stack?)
-    int32_t position = index + fZero;
-    U_ASSERT(position >= 0);
+    int32_t position;
+    if (uprv_add32_overflow(fZero, index, &position)) {
```

Comment:
This seems unnecessary. Instead I would
```
    U_ASSERT(0 <= index && index <= fLength);
    U_ASSERT(count <= (fLength - index));
```
... and then you need not add the UErrorCode.

Reply:
how can we be sure index + fZero won't overflow int32_t ?

Reply:
done